### PR TITLE
Maven-archetype for siddhi-testsuite

### PR DESCRIPTION
**Pull request by PrabodDunuwila**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: siddhi-testsuite/src/main/resources/archetype-resources/component/pom.xml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                <executions>
+                    <execution>
+                        <id>copy-jars</id>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>copy</goal>
+                        </goals>
+                        <configuration>
+                            <artifactItems>
+                                <!--Jars to be included for AppTest.-->
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/pom.xml]: Add more descriptive comment

Comment (pcnfernando) [reply siddhi-testsuite/src/main/resources/archetype-resources/component/pom.xml]: Say why you are bundling defaults

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
+package sample.containers;
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java]: License missing

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import java.io.IOException;
+import java.net.Socket;
+import java.net.URI;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+import static org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess;
+
+/**
+ * Siddhi Logger Service container class. This is a sample LoggerServiceContainer which can be used to log a message.
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java]: Rephrase the comments.

Sample container extending the Container framework which logs events received through an HTTP endpoint.

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ */
+public class LoggerServiceContainer extends GenericContainer<LoggerServiceContainer> {
+
+    private static final String IMAGE_NAME = "siddhi-logservice";
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java]: You have added a image built locally. Clean the image and push it to siddhiio docker repo

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            if (!isRunning()) {
+                throw new ContainerLaunchException("Logger Service  Container failed to start");
+            }
+            try (Socket s = new Socket("0.0.0.0", this.getMappedPort(DEFAULT_LOGSERVICE_PORT))) {
+                return null;
+            }
+        });
+    }
+
+    private HTTPClient.HTTPResponseMessage callHealthAPI() {
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/main/java/sample/containers/LoggerServiceContainer.java]: This method is never used

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
+package sample;
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java]: License missing

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Class for AbstractTemperatureAlertTests of Temp-Alert-App. This class includes the common logic for integration
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java]: Rephrase as Abstract tests for Temp-Alert-App. This class includes the common logic for integration tests and blackbox tests.

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    public NatsContainer natsContainer;
+    public SiddhiRunnerContainer siddhiRunnerContainer;
+    public NatsClient natsClient;
+    private String natsClusterId;
+    private String natsUrl;
+    private String natsInputDestination;
+    private String natsOutputDestination;
+
+    @BeforeClass
+    public abstract void setUpCluster() throws IOException, InterruptedException;
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java]: Add comments

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+ * Description: Used for temperature monitoring and anomaly detection. Consumes events from a Nats topic,
+ *              filters the event under types 'monitored' and 'internal'.
+ *              Monitored events are then sent through a pattern and the matched events will be alerted to a Nats topic.
+ *              The internal events are persisted to a table.
+ */
+public abstract class AbstractTemperatureAlertTests {
+
+    public WaitingConsumer siddhiLogConsumer = new WaitingConsumer();
+
+    public MySQLContainer mySQLContainer;
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java]: Why these containers are in Abstract class. These are only used in Integration tests?

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        natsClient.publish(natsInputDestination, "{\n" +
+                "    \"event\": {\n" +
+                "        \"type\": \"dummyType\",\n" +
+                "        \"deviceID\": \"dummyDeviceID\",\n" +
+                "        \"temp\": 50.0,\n" +
+                "        \"roomID\": \"dummyRoomID\"\n" +
+                "    }\n" +
+                "}");
+        try {
+            siddhiLogConsumer.waitUntil(frame ->
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/AbstractTemperatureAlertTests.java]: You cannot assert from logConsumer for Blackbox teste right? Update the Temperature app to publish the result to the log service and assert from there

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
+package sample.blackbox;
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java]: license missing

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class for black box testing of Temp-Alert-App. Run black-box testing by configuring the Siddhi Runner instances
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java]: Rephrase class comments as in integration tests

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * Description: Used for temperature monitoring and anomaly detection. Consumes events from a Nats topic,
+ *              filters the event under types 'monitored' and 'internal'.
+ *              Monitored events are then sent through a pattern and the matched events will be alerted to a Nats topic.
+ *              The internal events are persisted to a table.
+ */
+public class BlackBoxTestsOfTemperatureApp extends AbstractTemperatureAlertTests {
+    private static final Logger logger = LoggerFactory.getLogger(BlackBoxTestsOfTemperatureApp.class);
+
+    //Provide configurations for Nats and mysql environments.
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java]: Rephrase the comment as **Update the Nats and MySql configurations as per your environment for Siddhi test framework to communicate with the said instances**

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        Map<String, String> envMap = new HashMap<>();
+        envMap.put("CLUSTER_ID", NATS_CLUSTER_ID);
+        envMap.put("INPUT_DESTINATION", NATS_INPUT_DESTINATION);
+        envMap.put("OUTPUT_DESTINATION", NATS_OUTPUT_DESTINATION);
+        envMap.put("NATS_URL", NATS_BOOTSTRAP_URL);
+        envMap.put("DATABASE_URL", DATABSE_JDBC_URL);
+        envMap.put("USERNAME", DATABSE_USERNAME);
+        envMap.put("PASSWORD", DATABSE_PASSWORD);
+        envMap.put("JDBC_DRIVER_NAME", DATABSE_DRIVER_NAME);
+        siddhiRunnerContainer = new SiddhiRunnerContainer("siddhiio/siddhi-runner-alpine:latest-dev")
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/blackbox/BlackBoxTestsOfTemperatureApp.java]: Is it possible to get the Siddhi Runner image name and version during interactive mode and update the classes accordingly. Use **siddhiio/siddhi-runner-alpine:latest-dev** as defaults

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/unit/UnitTestsOfTemperatureApp.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+public class UnitTestsOfTemperatureApp {
+
+    private static final Logger logger = LoggerFactory.getLogger(UnitTestsOfTemperatureApp.class);
+    private static URL appUrl = Resources.getResource("artifacts/apps/Temp-Alert-App.siddhi");
+    private volatile AtomicInteger count = new AtomicInteger(0);
+
+    @BeforeClass
+    private void setUpTest() {
+        Map<String, String> envMap = new HashMap<>();
+        envMap.put("CLUSTER_ID", "");
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/sample/unit/UnitTestsOfTemperatureApp.java]: Check whether without setting the envs to ""; works?

**Review by pcnfernando — CHANGES_REQUESTED** [review]

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/resources/testng.xml [hunk 16]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+
+<suite name="Siddhi-Testsuite">
+    <test name="Siddhi-Testsuite-Tests" parallel="false">
+        <classes>
+            <class name="sample.blackbox.BlackBoxTestsOfTemperatureApp"/>
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/resources/testng.xml]: You cannot run Blackbox tests without user configuring the URLs. Comment the said class from testng and say steps to enable

File: README.md [hunk 17]
```diff
@@ -96,6 +96,24 @@ To generate specific type of extensions archetype, issue the command from your C
...
+           
+    | Properties | Description | Mandatory | Default Value |
+    | ------------- |-------------| ---- | ----- |
+    | _nameOfTestsuite | Name of TestSuite | Y | - 
+    | groupIdPostfix| Name of TestSuite is added as postfix to the groupId as a convention | N | {_nameOfTestsuite}
+    | artifactId | Artifact Id of the project | N | siddhi-testsuite-{_nameOfTestsuite}
+    | classNameOfAbstractTest | Class name of the AbstractTest logic common for both Integration and BlackBox tests | N | AbstractTestsOf{_nameOfTestsuite}
+    | classNameOfUnitTest | Class name of the UnitTest | N | UnitTestsOf{_nameOfTestsuite}
+    | classNameOfIntegrationTest | Class name of the IntegrationTest | N | IntegrationTestsOf{_nameOfTestsuite}
+    | classNameOfBlackBoxTest | Class name of the BlackBoxTest | N | BlackBoxTestsOf{_nameOfTestsuite}
```
Comment (niveathika) [inline README.md]: Better if we change the names of the classes to, `{_nameOfTestsuite}UnitTestCase`, similar changes to the rest...

File: siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/template/unit/__classNameOfUnitTest__.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * This is the class for unit testing. Siddhi is used in an embedded sandbox mode to make sure no network connections
+ * or external dependencies are required for testing. This requires to write a Java code to test the Siddhi app.
+ *
+ */
+public class ${classNameOfUnitTest} {
+
+    private static final Logger logger = LoggerFactory.getLogger(${classNameOfUnitTest}.class);
+
+    @BeforeClass
```
Comment (pcnfernando) [inline siddhi-testsuite/src/main/resources/archetype-resources/component/src/test/java/template/unit/__classNameOfUnitTest__.java:18]: Add comments to all empty bodies in the template

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**